### ISSUE TITLE
[feature] Add a hook for Jobs that allows them to run without Helm waiting for completion

### ISSUE THREAD

**Issue by rbren** [helm/helm#10555]
Carried over from https://github.com/helm/helm/issues/5891

Many charts kick of a Job upon install/upgrade - e.g. to run database migrations. These Jobs can take a long time to run.

Currently the helm CLI waits for the Job to complete successfully before exiting. If the Job takes hours, this usually means that Helm will timeout and consider the install/upgrade a failure.

Current workaround is to create a Job that `kubectl apply`s the Job you actually want, which is a pain.

The proposal here is to add a hook or similar mechanism to tell Helm not to wait for a particular Job to finish before declaring the install/upgrade successfull

**Comment by TBBle** [comment]
My specific suggestion is a new hook policy, `hook-execution-policy` (parallels [`hook-delete-policy`](https://helm.sh/docs/topics/charts_hooks/#hook-deletion-policies)) with the value `non-blocking` to specify that the outcome of a hook does not affect the result of chart deployment.

It probably only makes sense with the `post-*` hooks, since a `pre-*` hook is by definition needed to be run _before_ something in the Chart, and a `test` hook isn't run during chart install anyway, but with `helm test` where presumably you want the _outcome_ of the test. But I wouldn't block this being done, any more than I'd block this from being applied to other resources than Jobs. No use-case jumps out at me, but that doesn't mean there isn't one.

See https://github.com/helm/helm/issues/5891#issuecomment-705404292 for more details.

An alternative idea would be a different type of hook, but there's already a lot of hook types covering two axes, so I think this would be an undesirable complexity increase.

**Comment by Warxcell** [comment]
One thing that can be done immediately:

If Hook is a Job and job have `spec.suspend: true` don't wait for it, because it will never complete, as it will never start.

**Comment by thesuperzapper** [comment]
This is still an important feature.

**Comment by sudermanjr** [comment]
activity?

**Comment by sudermanjr** [comment]
Agreed, a good feature request still.

**Comment by matteogazzadi** [comment]
Basically impossible to have jobs with suspend:true and hook due to this.

**Comment by kyrylogy** [comment]
I want to convert CronJob to Job with with another post-hook Job, but post-hook Job waits for CronJob completion, which will happen only when the schedule time is reached. Looking for a solution to CronJob -> Job post install & this blocking problem.

**Comment by kladiv** [comment]
+1 really useful

**Comment by joejulian** [comment]
Happy to have a contribution to address this. Probably should start with a [HIP](https://github.com/helm/community/blob/main/hips/hip-0001.md) also take a look at the [contributing](https://github.com/helm/helm/blob/main/CONTRIBUTING.md) doc.

**Comment by scndel** [comment]
For us too, it would be really useful to have a **post hook** not blocking `helm upgrade`. 
Here's our use case to show something concrete about the need : after upgrading one instance, we need eventually to update some data with an external api. Depending on the amount of entries, the job duration can significantly vary : from seconds to minutes.

**Comment by RomanKrut** [comment]
Also interested in this, in my current project I need to run 10 and more different scripts in parallel, I`m using dynamic range in simple helm template for job to create many jobs with different args, but when some job is too long for execute, my deploy takes long time, which is not good.
Does some one have proper solution for helm to not wait jobs complete status?

**Comment by meysam81** [comment]
+1

**Comment by joejulian** [comment]
Please don't spam the comments. Just add your emoji to the OP.

**Comment by thesuperzapper** [comment]
@joejulian After thinking about this more, we can address this simply by creating a new `helm.sh/hook` that runs DURING "upgrade" and "install" rather than pre/post.

With this new hook, if a chart has a one-time Job that must run before the main Deployments will enter a "Ready" state (for example, creating DB schemas), the user can do the following:

1. Create a Job that does the one-time preparation:
    - Annotate with `helm.sh/hook: install,upgrade` (does not exist yet as of helm 3.11)
    - Annotate with `helm.sh/hook-delete-policy: before-hook-creation,hook-succeeded`
        - _NOTE: this is important as the `PodSpec` of a `Job` is immutable_
2. Use `init-containers` on their main Deployments which check for the completion of the one-time task

In this approach, even when the `--wait` flag is passed the Job will run, preventing deadlocks.

**Comment by oskarfightingthekube** [comment]
Any ETA for this feature?

**Comment by bakoontz2** [comment]
Has this been implemented yet?  Elsewhere in another issue it was suggested this had been implemented, but I'm not seeing it in 3.12:

manifest_sorter.go:192: info: skipping unknown hook: "install,upgrade"

Strangely, the hook that's "unknown" is still run, and the job I'm using this on actually no longer blocks.  So I'm a bit confused:  Are unknown hooks simply ignored?

**Comment by bakoontz2** [comment]
Why would this be marked "stale" if it's still an issue that has yet to be addressed?

**Comment by gjenkins8** [comment]
> Why would this be marked "stale" if it's still an issue that has yet to be addressed?

Stale represents lack of progress. Since Helm is a community driven project, that means the community contributing a proposal/solution (or generally making progress).

**Comment by bakoontz2** [comment]
Ah, so basically a way to keep the requested feature list auto-pruned.  Nice.

**Comment by gjenkins8** [comment]
> Ah, so basically a way to keep the requested feature list auto-pruned. Nice.

Correct. Gotta tame the backlog somehow.

The issue here can be reopened. But that would be from someones intention of driving the proposal forward (to see if it could be accepted and implemented). Otherwise, it would just get marked Stale and be auto-closed next month.

**Comment by PrinsonF77** [comment]
Is this issue still open? If so, I would like to take it up. I already have a plan of action which follows from @TBBle 's comment.
1) Create a new hook policy `hook-execution-policy` with any one of `blocking` or 'non-blocking` as the possible values (Defaults to blocking)
2) When `hook-execution-policy` is set to `non-blocking`, check if the hook is a `post-*` hook. If yes, skip the current flow, where helm waits for the hook to reach completion. 3 Questions arise here:
     - What happens to `LastRun.Phase` for this hook ? Should it be set to `HookPhaseRunning` or `HookPhaseUnknown`
     - What happens to `LastRun.CompletedAt` for this hook ?
     - The hook must be prevented from deletion under the `hook-succeeded` deletion policy. How do we achieve this?
3) When `hook-execution-policy` is set to `blocking`, go with the current flow, where helm waits for the hooks to complete

Any comments would be appreciated. I already have some changes ready and would like to refine them based on inputs

**Comment by thesuperzapper** [comment]
@gjenkins8 it's pretty clear that this is a very important issue with lots of interest, can we please reopen it?

If only for your own sake, so that people don't create more issues for the same thing.
